cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -179,6 +187,44 @@ class ChunkedArraySorter : public TypeVisitor {
     std::copy(temp_indices, temp_indices + (range_end - range_begin), range_begin);
   }
 
+  // Duplicate of ArrayCompareSorter
+  static Result<std::shared_ptr<Array>> RanksWithNulls(
+      const std::shared_ptr<Array>& array) {
+    // Notes:
+    // * The order is always ascending here, since the goal is to produce
+    //   an exactly-equivalent-order of the dictionary values.
+    // * We're going to re-emit nulls in the output, so we can just always consider
+    //   them "at the end".  Note that choosing AtStart would merely shift other
+    //   ranks by 1 if there are any nulls...
+    RankOptions rank_options(SortOrder::Ascending, NullPlacement::AtEnd,
+                             RankOptions::Dense);
+
+    auto data = array->data();
+    std::shared_ptr<Buffer> null_bitmap;
+    if (array->null_count() > 0) {
+      null_bitmap = array->null_bitmap();
+      data = array->data()->Copy();
+      if (data->offset > 0) {
+        ARROW_ASSIGN_OR_RAISE(
+            null_bitmap,
+            arrow::internal::CopyBitmap(arrow::default_memory_pool(), null_bitmap->data(),
+                                        data->offset, data->length));
+      }
+      data->buffers[0] = nullptr;
+      data->null_count = 0;
+    }
+    ARROW_ASSIGN_OR_RAISE(auto rank_datum,
+                          CallFunction("rank", {std::move(data)}, &rank_options));
```

Comment:
I think using rank here is expensive and perhaps unnecessary? IIUC The rank is only used to compare two elements during merge sorting. Can we compare them by their values directly? Since `rank` internally also sorts the array, we are essentially sorting it twice.